src/settings.cpp

```diff
@@ -1709,20 +1698,65 @@ void IntSettingDesc::ChangeValue(const void *object, int32 newval) const
 	if (_save_config) SaveToConfig();
 }
 
+/**
+ * Given a name of setting, return a company setting description of it.
+ * @param name  Name of the company setting to return a setting description of
+ * @return Pointer to the setting description of setting \a name if it can be found,
+ *         \c nullptr indicates failure to obtain the description.
+ */
+const SettingDesc *GetCompanySettingFromName(const char *name)
+{
+	for (auto &sd : _company_settings) {
+		if (!SlIsObjectCurrentlyValid(sd->save.version_from, sd->save.version_to)) continue;
+		if (strcmp(sd->name, name) == 0) return sd.get();
+	}
+
+	return nullptr;
+}
+
+/**
+ * Given a name of setting, return a setting description of it.
+ * @param name  Name of the setting to return a setting description of
+ * @return Pointer to the setting description of setting \a name if it can be found,
+ *         \c nullptr indicates failure to obtain the description.
+ */
+const SettingDesc *GetSettingFromName(const char *name)
+{
+	/* First check all full names */
+	for (auto &sd : _settings) {
+		if (!SlIsObjectCurrentlyValid(sd->save.version_from, sd->save.version_to)) continue;
+		if (strcmp(sd->name, name) == 0) return sd.get();
+	}
+
+	/* Then check the shortcut variant of the name. */
+	for (auto &sd : _settings) {
+		if (!SlIsObjectCurrentlyValid(sd->save.version_from, sd->save.version_to)) continue;
+		const char *short_name = strchr(sd->name, '.');
+		if (short_name != nullptr) {
+			short_name++;
+			if (strcmp(short_name, name) == 0) return sd.get();
+		}
+	}
+
+	if (strncmp(name, "company.", 8) == 0) name += 8;
+	/* And finally the company-based settings */
+	return GetCompanySettingFromName(name);
+}
+
 /**
  * Network-safe changing of settings (server-only).
  * @param tile unused
  * @param flags operation to perform
- * @param p1 the index of the setting in the SettingDesc array which identifies it
+ * @param p1 unused
  * @param p2 the new value for the setting
  * The new value is properly clamped to its minimum/maximum when setting
- * @param text unused
+ * @param text the name of the setting to change
  * @return the cost of this operation or an error
  * @see _settings
  */
 CommandCost CmdChangeSetting(TileIndex tile, DoCommandFlag flags, uint32 p1, uint32 p2, const char *text)
 {
-	const SettingDesc *sd = GetSettingDescription(p1);
```

Comment:
Is text guaranteed to be non-`nullptr`? If not, `strcmp` on `nullptr` is undefined behaviour. 
Yes, from how we call it, but I'm rather thinking about a bad actor trying to mess things up.